tripal/tripal
# Compatibility with Drupal 10.0 🎉  (and keeping 9.4 and 9.5)

# Tripal 4 Core Dev Task

Issue #1510 and #1378

### Tripal Version: 4

## Description

This PR achieves compatibility of Tripal 4 with Drupal 10.0 while keeping compatibility with Drupal 9.4.x and 9.5.x. We will be dropping support for 9.2.x and 9.3.x which are no longer supported by Drupal anyway.

### Summary of Changes

**As this was a combined effort between myself and Valentin, I'll tag the person responsible in the list below.**

- Cloning functions originally installed during Tripal Chado module install are not added to a temporary schema as needed.  (Valentin)
    - This ensures that the functionality is available even if you choose to install Drupal in a separate database than Chado. 
    - It also ensures we do not clutter the public schema with functions we may never use and 
    - removes the assumption that Drupal is stored in the public schema.
- Updates all automated tests to use the `stark` theme instead of the `stable` testing theme.  (Lacey)
    - Stable testing theme removed in Drupal 10. Choices are now stark or Olivia (for testing expecting HTML classes)
- Drupal 10 moved the PostgreSQL database connection and schema classes into a different namespace. This PR adopts the new namespace.  (Valentin)
    - Was \Drupal\Core\Database\Driver\pgsql\Connection
    - Now Drupal\pgsql\Driver\Database\pgsql\Connection
- Updated Jobs view page to use the new RouteObjectInterface as \Symfony\Cmf\Component\Routing\RouteObjectInterface was removed in Drupal 10.  (Lacey)
- Updated Regex for Schema names to allow matching within a string (removed `^` and `$`).   (Valentin)
- Also added a Table names regex to make it more clear what we are matching and provide support for different regex in the future as needed.  (Valentin)
- Updated TripalDbx->getDrupalSchemaName() to remove assumption that Drupal was installed in a schema named public and to be less brittle in the testing environment when checking if the database is postgresql.  (Valentin + Lacey)
- Combined TripalDbxConnections schemaName and extraSchemas into a single usedSchemas array.  (Valentin)
- Fix bug in TripalDbxConnection constructor which was making incorrect changes to the settings.yml connection options and caused errors in Drupal 10.  (Lacey + Valentin)
- Updated getSchemaName() and getQuotedSchemaName() to match expectations if no schema was specified.  (Lacey)
- Remove setPrefix() method from TripalDbxConnection which is no longer needed.  (Valentin)
- Complete rewrite of prefixTables() in TripalDbxConnection for better handling of many cases and support for changes made in Drupal 10.  (Valentin + Lacey)
- Update automated tests to match PHPUnit 10 best practices for mock objects.  (Lacey)
- Updated a number of tests which were out of sync with the codebase and thus failing to match current expectations.  (Lacey)
- Fixed bug where the Default schema was not being set automatically. Now the first chado installed becomes the default schema. This can still be changed in the Chado management tools.  (Lacey)
- Update cloning Task to use the new temporary cloning function.  (Lacey)


## Testing?

### Automated Testing

Automated testing should pass for all workflows testing Drupal 9.4.x, 9.5.x and 10.0.x. Support is dropped for 9.3.x and 9.2.x so those workflows are expected to fail.

### Docker for this branch

```
git clone https://github.com/tripal/tripal.git tripal-1517-d10
cd tripal-1517-d10
git checkout tv4g2-issue-1378-tripaldbx-d10-compatibility
docker build --tag=tripaldocker:testD10 --build-arg drupalversion="10.0.x-dev" --build-arg chadoschema=testchado --file tripaldocker/Dockerfile-php8.1-pgsql13 ./
docker run --publish=80:80 --name=testD10 -tid --volume=`pwd`:/var/www/drupal9/web/modules/contrib/tripal tripaldocker:testD10
docker exec testD10 service postgresql restart
```

For more reasons on why the above commands check out @dsenalik documentation here: https://tripaldoc.readthedocs.io/en/latest/contributing/docker-for-testing.html#testing-an-unmerged-pull-request

### Manual testing

Confirm that you can 
- install a new Tripal site (also confirmed by docker build),
- Create Tripal content types
- Create Tripal Content
- Anything else that tickles your fancy that you were able to do in Tripal 4 before should still work.

NOTE: Data importers are not quite working in Drupal 10 just yet. This will be fixed in a future PR.
